src/Inventory/Asset/OperatingSystem.php

```diff
@@ -82,22 +85,34 @@ public function prepare(): array
             $val->install_date = date('Y-m-d', strtotime($val->install_date));
         }
 
-        if (
-            property_exists($val, 'operatingsystemarchitectures_id')
-            && $val->operatingsystemarchitectures_id != ''
-        ) {
-            $rulecollection = new RuleDictionnaryOperatingSystemArchitectureCollection();
-            $res_rule = $rulecollection->processAllRules(['name' => $val->operatingsystemarchitectures_id]);
+        $mapping = [
+            'operatingsystems_id'               => RuleDictionnaryOperatingSystemCollection::class,
+            'operatingsystemversions_id'        => RuleDictionnaryOperatingSystemVersionCollection::class,
+            'operatingsystemservicepacks_id'    => RuleDictionnaryOperatingSystemServicePackCollection::class,
+            'operatingsystemarchitectures_id'   => RuleDictionnaryOperatingSystemArchitectureCollection::class,
+        ];
+
+        //copy inventory data to prevent data override
+        $inventory_val = $val;
+        foreach ($mapping as $key => $rule_class) {
+            $rulecollection = new $rule_class();
+            $rule_input = [
+                'name'              => property_exists($inventory_val, $key) ? $inventory_val->{$key} : "",
+                'os_name'           => $inventory_val->operatingsystems_id ?? '',
+                'os_version_name'   => $inventory_val->operatingsystemversions_id ?? '',
+                'servicepack_name'  => $inventory_val->operatingsystemservicepacks_id ?? '',
+                'arch_name'         => $inventory_val->operatingsystemarchitectures_id ?? '',
+            ];
```

Comment:
You could define the rule input directly outside the loop
```suggestion
        $rule_input = [
            'name'              => property_exists($val, $key) ? $val->{$key} : "",
            'os_name'           => $val->operatingsystems_id ?? '',
            'os_version_name'   => $val->operatingsystemversions_id ?? '',
            'servicepack_name'  => $val->operatingsystemservicepacks_id ?? '',
            'arch_name'         => $val->operatingsystemarchitectures_id ?? '',
        ];
        foreach ($mapping as $key => $rule_class) {
            $rulecollection = new $rule_class();
```

Reply:
I don't think so ```name``` changed at each iteration

Reply:
done